shipment_advice_cash_on_delivery/models/stock_picking.py

```diff
@@ -0,0 +1,45 @@
+# Copyright 2018 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
+# Copyright 2023 ACSONE SA/NV
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import fields, models
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    cash_on_delivery_invoice_ids = fields.Many2many(
+        "account.move", string="COD Invoices", copy=False, readonly=True
+    )
+
+    def _invoice_at_shipping(self):
+        self.ensure_one()
+        res = super()._invoice_at_shipping()
+        res = res or self.sale_id.payment_mode_id.cash_on_delivery
+        return res
+
+    def _invoicing_at_shipping_validation(self, invoices):
+        # COD invoices will be validated automatically
+        cod_invoices_to_validate = invoices.filtered(
+            lambda invoice: invoice.payment_mode_id.cash_on_delivery
+            and invoice.payment_mode_id.auto_validate_invoice
+        )
+        # Non-COD invoices need to check in the call of `super`
+        invoices_to_validate = invoices.filtered(
+            lambda invoice: not invoice.payment_mode_id.cash_on_delivery
+        )
+        res = super()._invoicing_at_shipping_validation(invoices_to_validate)
```

Comment:
Don't you need to do this to also validate the cod invoices to validate?
```suggestion
        res = super()._invoicing_at_shipping_validation(cod_invoices_to_validate | invoices_to_validate)
```

Then you can just return res

Reply:
@jbaudoux I don't get your point,  because in [here](https://github.com/OCA/account-invoicing/pull/1695#discussion_r1547661277):
- if the COD invoices with payment mode is COD and auto validate, it should be return without calling `super()`
- and the others will be passed to the `super()` to check

Reply:
indeed, I though this was doing the validation but you're right this is only computing if this requires validation